onnx/defs/quantization/defs.cc

```diff
@@ -15,6 +15,68 @@ For saturation, it saturates to [0, 255] if it's uint8, or [-128, 127] if it's i
 For (x / y_scale), it's rounding to nearest ties to even. Refer to https://en.wikipedia.org/wiki/Rounding for details. 'y_zero_point' and 'y' must have same type.
 )DOC";
 
+bool BuildContextDependentFunctionBodyQuantizeLinear(
+    const FunctionBodyBuildContext& ctx,
+    const OpSchema& schema,
+    FunctionProto& functionProto) {
+  FunctionBuilder builder(functionProto);
+
+  auto* input_type = ctx.getInputType(0);
+  if (!input_type->has_tensor_type()) {
+    fail_schema("Invalid x input type. Expected a tensor type");
+  }
+  auto input_elt_type = input_type->tensor_type().elem_type();
+
+  auto output_elt_type = TensorProto::UINT8;
+  float quantized_min = 0, quantized_max = 255;
+
+  if (ctx.hasInput(2)) {
+    auto* zero_point_type = ctx.getInputType(2);
+    if (!zero_point_type->has_tensor_type()) {
+      fail_schema("Invalid y_zero_point input type. Expected a tensor type");
+    }
+    if (zero_point_type->tensor_type().elem_type() == TensorProto::UINT8) {
+      quantized_min = 0;
+      quantized_max = 255;
+      output_elt_type = TensorProto::UINT8;
+    } else if (zero_point_type->tensor_type().elem_type() == TensorProto::INT8) {
+      quantized_min = -128;
+      quantized_max = 127;
+      output_elt_type = TensorProto::INT8;
+    } else {
+      fail_schema("Invalid zero_point_type. Expected TensorProto::UINT8 or TensorProto::INT8");
+    }
+  }
+
+  bool is_per_axis = ctx.getInputType(1)->tensor_type().shape().dim().size() == 1 &&
+      ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value() > 1;
+
+  builder.Const("quantized_min", ToTensor(quantized_min));
+  builder.Const("quantized_max", ToTensor(quantized_max));
+  if (is_per_axis) {
+    auto* axis_attribute = ctx.getAttribute("axis");
+    int axis = axis_attribute ? ctx.getAttribute("axis")->i() : 1;
+    int x_rank = ctx.getInputType(0)->tensor_type().shape().dim().size();
+    std::vector<int64_t> y_scale_shape(x_rank - axis, 1);
+    y_scale_shape[0] = ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value();
+    builder.Const("y_scale_shape", y_scale_shape);
+    builder.Add("y_scale_reshape = Reshape (y_scale, y_scale_shape)");
+    builder.Add("x_s = Div (x, y_scale_reshape)");
+    builder.Add("y_zero_point_reshape = Reshape (y_zero_point, y_scale_shape)");
+    builder.Add("y_zero_point_cast = Cast(y_zero_point_reshape)", "to", (int64_t)(input_elt_type));
+  } else {
+    builder.Add("x_s = Div (x, y_scale)");
+    builder.Add("y_zero_point_cast = Cast(y_zero_point)", "to", (int64_t)(input_elt_type));
```

Comment:
One trouble with Quantization is that it mentions the use of saturated arithmetic. Will cast do the right thing?

Reply:
cast from quantized (narrow) to dequantized (wide) is ok. To cast from dequantized to quantized, we need to clip first.